File: backend/src/main/java/dev/arc/engine/Engine.java
package dev.arc.engine;

import dev.arc.api.ArcException;
import dev.arc.model.Definition;
import dev.arc.model.Definition.*;
import dev.arc.source.*;
import java.util.*;
import org.springframework.stereotype.Component;

@Component
public class Engine {
  public record Step(
      String ruleId,
      Integer version,
      String nodeId,
      String label,
      String type,
      Object value,
      String branch,
      int depth) {}

  public record Result(
      Object result, List<Step> trace, long durationMicros, List<Parameters.Read> sources) {}

  private record Value(Object value, String origin) {}

  private final Validator validator;

  public Engine(Validator validator) {
    this.validator = validator;
  }

  public Result execute(
      String ruleId,
      Integer version,
      Definition definition,
      Map<String, Object> inputs,
      RuleResolver resolver) {
    return execute(ruleId, version, definition, inputs, resolver, new Parameters(null));
  }

  public Result execute(
      String ruleId,
      Integer version,
      Definition definition,
      Map<String, Object> inputs,
      RuleResolver resolver,
      Parameters parameters) {
    long start = System.nanoTime();
    List<Step> trace = new ArrayList<>();
    Object value =
        run(ruleId, version, definition, inputs, resolver, trace, new HashSet<>(), 0, parameters);
    return new Result(value, trace, (System.nanoTime() - start) / 1000, parameters.reads());
  }

  private Object run(
      String ruleId,
      Integer version,
      Definition d,
      Map<String, Object> inputs,
      RuleResolver resolver,
      List<Step> trace,
      Set<String> stack,
      int depth,
      Parameters parameters) {
    if (depth > 16) throw ArcException.invalid("Rule nesting exceeds 16 levels");
    String key = ruleId + "@" + version;
    if (!stack.add(key)) throw ArcException.invalid("Circular rule reference: " + key);
    try {
      validator.validate(d, resolver);
      var plan = new GraphPlan(d);
      Node input =
          plan.order.stream().filter(n -> n.type().equals("INPUT")).findFirst().orElseThrow();
      Map<String, Object> provided;
      try {
        provided = parameters.resolve(d.inputs(), inputs);
      } catch (ArcException e) {
        throw e.atNode(ruleId, version, input.id(), input.label());
      }
      Map<String, Map<String, Value>> scopes = new HashMap<>();
      Map<String, String> branches = new HashMap<>();
      Map<String, Object> outputs = new LinkedHashMap<>();
      // All parents have been resolved before a node runs, including skipped
      // branches. Fan-out gets independent scopes and a shared join runs once.
      for (Node node : plan.order) {
        Map<String, Value> values = new LinkedHashMap<>();
        boolean active = node.type().equals("INPUT");
        try {
          if (active)
            provided.forEach((name, value) -> values.put(name, new Value(value, input.id())));
          Map<String, Map<String, Value>> candidates = new LinkedHashMap<>();
          for (Edge edge : plan.incoming.getOrDefault(node.id(), List.of())) {
            if (!edge.sourceHandle().equals(branches.get(edge.source()))) continue;
            active = true;
            scopes
                .get(edge.source())
                .forEach(
                    (name, value) ->
                        candidates
                            .computeIfAbsent(name, k -> new LinkedHashMap<>())
                            .put(value.origin(), value));
          }
          if (!active) continue;
          for (var candidate : candidates.entrySet()) {
            var origins = candidate.getValue();
            var latest =
                origins.values().stream()
                    .filter(
                        v ->
                            origins.keySet().stream()
                                .noneMatch(other -> plan.ancestors.get(other).contains(v.origin())))
                    .toList();
            if (latest.size() != 1)
              throw ArcException.invalid(
                  "Conflicting upstream values for '"
                      + candidate.getKey()
                      + "'; use distinct result variable names before merging");
            values.put(candidate.getKey(), latest.getFirst());
          }
          if (trace.size() >= 1000) throw ArcException.invalid("Execution exceeds 1,000 steps");
          Map<String, Object> scope = new LinkedHashMap<>();
          values.forEach((name, value) -> scope.put(name, value.value()));
          Object value = null;
          String branch = "next";
          switch (node.type()) {
            case "INPUT" -> value = new LinkedHashMap<>(scope);
            case "FORMULA" -> value = Expressions.evaluate(node.expression(), scope);
            case "CONDITION" -> {
              value = Expressions.bool(Expressions.evaluate(node.expression(), scope));
              branch = value.toString();
            }
            case "REFERENCE" -> {
              var bound = new LinkedHashMap<String, Object>();
              if (node.bindings() != null)
                node.bindings()
                    .forEach((k, expr) -> bound.put(k, Expressions.evaluate(expr, scope)));
              value =
                  run(
                      node.ruleId(),
                      node.version(),
                      resolver.resolve(node.ruleId(), node.version()),
                      bound,
                      resolver,
                      trace,
                      stack,
                      depth + 1,
                      parameters);
            }
            case "OUTPUT" -> {
              value = Expressions.evaluate(node.expression(), scope);
              outputs.put(node.id(), value);
              branch = null;
            }
            default -> throw ArcException.invalid("Unknown node type");
          }
          if (node.type().equals("FORMULA") || node.type().equals("REFERENCE"))
            values.put(node.output(), new Value(value, node.id()));
          scopes.put(node.id(), values);
          branches.put(node.id(), branch);
          if (trace.size() >= 1000) throw ArcException.invalid("Execution exceeds 1,000 steps");
          trace.add(
              new Step(
                  ruleId, version, node.id(), node.label(), node.type(), value, branch, depth));
        } catch (ArcException e) {
          throw e.atNode(ruleId, version, node.id(), node.label());
        }
      }
      if (outputs.isEmpty()) throw ArcException.invalid("Execution did not reach an Output node");
      return outputs.size() == 1 ? outputs.values().iterator().next() : outputs;
    } catch (ArcException e) {
      throw e.inRule(ruleId, version);
    } finally {
      stack.remove(key);
    }
  }
}

File: backend-rails/lib/arc/engine.rb
module Arc
  class Engine
    def initialize(resolver, sources)
      @resolver, @sources, @validator = resolver, sources, Validator.new
    end

    def execute(rule_id, version, definition, inputs)
      started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      @trace, @stack, @parameters = [], Set.new, Parameters.new(@sources)
      result = run(rule_id, version, definition, inputs, 0)
      { "ruleId" => rule_id, "version" => version, "result" => result, "trace" => @trace,
        "durationMicros" => ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - started) * 1_000_000).to_i, "sources" => @parameters.reads }
    end

    def run(rule_id, version, d, inputs, depth)
      fail_rule("Rule nesting exceeds 16 levels") if depth > 16
      key = "#{rule_id}@#{version || 'null'}"
      fail_rule("Circular rule reference: #{key}") unless @stack.add?(key)
      begin
        @validator.validate(d, @resolver)
        plan = GraphPlan.new(d)
        input = plan.order.find { |n| n["type"] == "INPUT" }
        begin
          provided = @parameters.resolve(d["inputs"], inputs)
        rescue ArcError => e
          raise e.at_node(input, rule_id: rule_id, version: version)
        end
        scopes, branches, outputs = {}, {}, {}
        plan.order.each do |node|
          begin
            active = node["type"] == "INPUT"
            values = active ? provided.to_h { |k, v| [k, [v, input["id"]]] } : {}
            candidates = {}
            plan.incoming[node["id"]].each do |edge|
              next unless edge["sourceHandle"] == branches[edge["source"]]
              active = true
              scopes.fetch(edge["source"]).each { |name, value| (candidates[name] ||= {})[value[1]] = value }
            end
            next unless active
            candidates.each do |name, origins|
              latest = origins.values.reject { |v| origins.keys.any? { |other| plan.ancestors[other].include?(v[1]) } }
              fail_rule("Conflicting upstream values for '#{name}'; use distinct result variable names before merging") unless latest.size == 1
              values[name] = latest.first
            end
            fail_rule("Execution exceeds 1,000 steps") if @trace.size >= 1000
            scope = values.transform_values(&:first)
            branch = "next"
            value = case node["type"]
            when "INPUT" then scope.dup
            when "FORMULA" then Expressions.evaluate(node["expression"], scope)
            when "CONDITION"
              condition = Expressions.bool(Expressions.evaluate(node["expression"], scope))
              branch = condition.to_s
              condition
            when "REFERENCE"
              bound = (node["bindings"] || {}).transform_values { |expr| Expressions.evaluate(expr, scope) }
              run(node["ruleId"], node["version"], @resolver.call(node["ruleId"], node["version"]), bound, depth + 1)
            when "OUTPUT"
              branch = nil
              outputs[node["id"]] = Expressions.evaluate(node["expression"], scope)
            else fail_rule("Unknown node type")
            end
            values[node["output"]] = [value, node["id"]] if %w[FORMULA REFERENCE].include?(node["type"])
            scopes[node["id"]], branches[node["id"]] = values, branch
            fail_rule("Execution exceeds 1,000 steps") if @trace.size >= 1000
            @trace << { "ruleId" => rule_id, "version" => version, "nodeId" => node["id"], "label" => node["label"], "type" => node["type"], "value" => value, "branch" => branch, "depth" => depth }
          rescue ArcError => e
            raise e.at_node(node, rule_id: rule_id, version: version)
          end
        end
        fail_rule("Execution did not reach an Output node") if outputs.empty?
        outputs.size == 1 ? outputs.values.first : outputs
      rescue ArcError => e
        raise e.in_rule(rule_id, version)
      ensure
        @stack.delete(key)
      end
    end

    def fail_rule(message)
      raise ArcError.new(422, message)
    end
  end
end
